### PR TITLE
Fixes gh-4303 #4304

### LINKED ISSUE #4303: Register cron expression based on timezone of SCDF Server.
Problem description:

The timezone between Cloud Platform's cron expression and Spring Cloud Data Flow Server's cron expression does not match.

Solution description:
In the k8s configMap, specify the timezone of the k8s platform. And when registering the schedule, modify the cron expression by referring to the timezone of the k8s platform.

apiVersion: v1
kind: ConfigMap
metadata:
  name: scdf-server
  labels:
    app: scdf-server
data:
  application.yaml: |-
    management:
      metrics:
        export:
          prometheus:
            enabled: true
            rsocket:
              enabled: true
              host: prometheus-proxy
              port: 7001
    spring:
      cloud:
        dataflow:
          metrics.dashboard:
            url: 'https://grafana:3000'
          task:
            platform:
              **timezone: UTC**
              kubernetes:
                accounts:
                  default:
                    limits:
                      memory: 1024Mi
      datasource:
        url: jdbc:mysql://${MYSQL_SERVICE_HOST}:${MYSQL_SERVICE_PORT}/mysql
        username: root
        password: ${mysql-root-password}
        driverClassName: org.mariadb.jdbc.Driver
        testOnBorrow: true
        validationQuery: "SELECT 1"

Description of alternatives:
N/A

Additional context:
I have already modified the code. We'll share your edits through Pull Requests.

### PR DESCRIPTION
**Pull request by 23king**
Fixes #4303
add TaskPlatformConfigurationProperties.java (set platform timezone)
add cronutils dependency (9.1.3)

If the timzone of the platform exists, the cron expression is changed to the cron expression of the platform.
and
This changes the cron expression of the platform to the cron expression of the server even when importing the task.

@sabbyanandan 
Check this commit https://github.com/spring-cloud/spring-cloud-dataflow/pull/4304/commits/627d8c04b69e132ed8b448102bf7ca8156b186f4

### CODE REVIEW

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultSchedulerService.java [hunk 1]
```diff
@@ -219,24 +254,53 @@ public void schedule(String scheduleName, String taskDefinitionName, Map<String,
...
-		this.auditRecordService.populateAndSaveAuditRecordUsingMapData(AuditOperationType.SCHEDULE, AuditActionType.CREATE,
-				scheduleRequest.getScheduleName(), this.auditServiceUtils.convertScheduleRequestToAuditData(scheduleRequest),
+		this.auditRecordService.populateAndSaveAuditRecordUsingMapData(AuditOperationType.SCHEDULE,
+				AuditActionType.CREATE,
+				scheduleRequest.getScheduleName(),
+				this.auditServiceUtils.convertScheduleRequestToAuditData(scheduleRequest),
 				launcher.getName());
 	}
 
+	private static String convertCronExpressionForPlatformTimeZone(String cronExpression, ZoneId platformTimezone) {
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultSchedulerService.java:280]: Will this work with Cloud Foundries implementation of Cron Expressions which is based on Quartz

File: spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/StreamOperations.java [hunk 2]
```diff
@@ -161,4 +161,19 @@ void updateStream(String streamName, String releaseName, PackageIdentifier packa
 	 * @param properties scale deployment properties.
 	 */
 	void scaleApplicationInstances(String streamName, String appName, Integer count, Map<String, String> properties);
+
+	/**
+	 * Retrieves all the applications' logs for the given stream name
+	 * @param streamName name of stream for which to get logs
+	 * @return logs of said stream
+	 */
+	String streamExecutionLog(String streamName);
```
Comment (cppwfs) [inline spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/StreamOperations.java:170]: This Stream log changes need to be in a different PR.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultSchedulerService.java (hunk 1 above)
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultSchedulerService.java:280]: Will this conversion with with Cloud Foundry's cron expressions which is based on quartz?

File: spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/StreamOperations.java (hunk 2 above)
Comment (cppwfs) [inline spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/StreamOperations.java:170]: The Stream log changes need to be in a different PR.   The do not pertain to the current PR.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultSchedulerService.java [hunk 3]
```diff
@@ -182,13 +211,16 @@ public void schedule(String scheduleName, String taskDefinitionName, Map<String,
 			taskDefinition = new TaskDefinition(taskDefinition.getName(),
 					TaskServiceUtils.createComposedTaskDefinition(
 							taskNode.toExecutableDSL()));
-			taskDeploymentProperties = TaskServiceUtils.establishComposedTaskProperties(taskDeploymentProperties, taskNode);
+			taskDeploymentProperties = TaskServiceUtils.establishComposedTaskProperties(taskDeploymentProperties,
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultSchedulerService.java:214]: Looks like your formatter has a max limit per line.   Please remove that restriction.   Thanks.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/TaskConfigurationProperties.java [hunk 4]
```diff
@@ -77,6 +84,17 @@ public void setComposedTaskRunnerUri(String composedTaskRunnerUri) {
 		}
 	}
 
+	public ZoneId getPlatformTimezone() {
+		try {
+			return taskPlatformConfigurationProperties.getTimezone() == null ? null
+					: ZoneId.of(taskPlatformConfigurationProperties.getTimezone());
+		}
+		catch (ZoneRulesException ex) {
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/TaskConfigurationProperties.java:92]: Why are we logging this exception?   Seems like it would give us a false "null" value.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/configuration/JobDependencies.java [hunk 5]
```diff
@@ -83,16 +83,7 @@
...
-import org.springframework.cloud.dataflow.server.service.impl.DefaultLauncherService;
-import org.springframework.cloud.dataflow.server.service.impl.DefaultTaskDeleteService;
-import org.springframework.cloud.dataflow.server.service.impl.DefaultTaskExecutionInfoService;
-import org.springframework.cloud.dataflow.server.service.impl.DefaultTaskExecutionRepositoryService;
-import org.springframework.cloud.dataflow.server.service.impl.DefaultTaskExecutionService;
-import org.springframework.cloud.dataflow.server.service.impl.DefaultTaskJobService;
-import org.springframework.cloud.dataflow.server.service.impl.DefaultTaskSaveService;
-import org.springframework.cloud.dataflow.server.service.impl.TaskAppDeploymentRequestCreator;
-import org.springframework.cloud.dataflow.server.service.impl.TaskConfigurationProperties;
+import org.springframework.cloud.dataflow.server.service.impl.*;
```
Comment (cppwfs) [inline spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/configuration/JobDependencies.java:86]: Please resolve checkstyle issues for a clean compile.

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/TaskConfigurationProperties.java (hunk 4 above)
Comment (23king) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/TaskConfigurationProperties.java:92]: This is to ensure that the server runs without problems when the timezone is entered incorrectly, but the corresponding part will be deleted.

File: spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/StreamOperations.java (hunk 2 above)
Comment (23king) [reply spring-cloud-dataflow-rest-client/src/main/java/org/springframework/cloud/dataflow/rest/client/StreamOperations.java:170]: OK

File: spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultSchedulerService.java (hunk 1 above)
Comment (23king) [reply spring-cloud-dataflow-server-core/src/main/java/org/springframework/cloud/dataflow/server/service/impl/DefaultSchedulerService.java:280]: Even if the cron expression is quartz type, there is no problem.

File: spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/configuration/JobDependencies.java (hunk 5 above)
Comment (23king) [reply spring-cloud-dataflow-server-core/src/test/java/org/springframework/cloud/dataflow/server/configuration/JobDependencies.java:86]: OK

**Comment by cppwfs** [comment]
Sorry for the delay.  I want to discuss some of the nuances with this PR and story with project leads before proceeding.   I will get back to you as soon as I have had these discussions.

**Comment by 23king** [comment]
OK, Thank you

**Comment by cppwfs** [comment]
@23king Thank you for your contribution.   We do appreciate it.   However, we had a discussion on this topic today, and decided that do not wish to convert the schedule cron expression for each platform.   But rather store this timezone information for each platform.    The issue is located here: https://github.com/spring-cloud/spring-cloud-dataflow/issues/4341

**Comment by cppwfs** [comment]
Well after another discussion, folks changed their minds :-D

**Comment by cppwfs** [comment]
Closing this PR.  The solution enumerated in  https://github.com/spring-cloud/spring-cloud-dataflow/issues/4476 is the preferred method.

Thank you for your contribution.
